### PR TITLE
Fix burr() distribution stats() method and add other methods

### PR DESCRIPTION
**Pull request by rlucas7**
closes scipy issue #9544 (mean and stats method of burr() return erroneous results).

Also fixes the issues in fisk and burr12 that inherit from burr.

Adds unit test(s) for c<1 case (mean/variance are np.nan and c>1 case (mean is finite).

Also cleans up the docstring for burr and adds refs.

Retry of PR #9584 after rebase fail. See that PR for some discussion and comments, mostly but not entirely addressed. 

https://github.com/scipy/scipy/pull/9584

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
         # burr.pdf(x, c, d) = c * d * x**(-c-1) * (1+x**(-c))**(-d-1)
         return c * d * (x**(-c - 1.0)) * ((1 + x**(-c))**(-d - 1.0))
 
+    def _logpdf(self, x, c, d):
+        return (np.log(c) + np.log(d) + (-c - 1.0) * np.log(x) +
+                 (-d - 1.0) * np.log1p(x**(-c)))
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: `xlogy(-c-1, x)`

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
...
+
     def _cdf(self, x, c, d):
         return (1 + x**(-c))**(-d)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Maybe route this through logcdf

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
...
 
+    def _logcdf(self, x, c, d):
+        return -d * np.log1p(x**(-c))
+
+    def _sf(self, x, c, d):
+        return 1 - (1 + x**(-c))**(-d)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Maybe add _logsf using log1p, too?

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
...
+        e1 = self._munp(1, c, d)
+        e2 = self._munp(2, c, d)
+        e3 = self._munp(3, c, d)
+        e4 = self._munp(4, c, d)
+        mu = np.where(c > 1.0, e1, np.nan)
+        mu2_if_c = e2 - mu**2
+        mu2 = np.where(c > 2.0, mu2_if_c, np.nan)
+        g1 = _lazywhere(c > 3.0, (c, e1, e2, e3, mu2_if_c),
+             lambda c, e1, e2, e3, mu2_if_c:
+             (e3 - 3 * e2 * e1 + 2 * e1**3) / np.sqrt((mu2_if_c)**3),
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Nit: I feel grouping operators by priority a bit more readable. Like so, `e3 - 3*e2*e1 + ...`. No need to change it here though.

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -3450,6 +3450,18 @@ def test_burr12_ppf_small_arg():
...
+def test_burr_nan_mean_var():
+    c, d = 0.5, 3
+    mean, variance = stats.burr(c, d).stats()
+    assert_(np.isnan(mean))
+    assert_(np.isnan(variance))
+
+def test_burr_stats():
+    c, d = 5.0, 3
+    mean, variance = stats.burr(c, d).stats()
+    mean_hc, variance_hc = 1.4110263183925857, 0.22879948026191643
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py]: An obligatory question about hardcoded numbers: how to reproduce them? IOW, if this test starts failing in the future, how does a future maintainer check these numbers?

File: scipy/stats/tests/test_distributions.py [hunk 6]
```diff
@@ -3450,6 +3450,18 @@ def test_burr12_ppf_small_arg():
...
+    assert_allclose(mean_hc, mean_hc)
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py]: Typo,  `mean, mean_hc`

**Review by ev-br — COMMENTED** [review]
LGTM modulo nitpicks

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: These are the raw moments, e.g. expected value of `x`, `E(X) = e1` second raw moment of `x`, `E(X^2) = e2` etc.
Do I understand correctly that your recommendation is to do the calculations directly on the  mean, variance, skewness, and kurtosis? I think that would be more code but I could do it if needed, would a comment here defining `e1`, `e2` etc help readability? Something like
```
#ek is the kth raw moment, e.g. e1 is the mean e2-e1**2 is variance etc.  
```

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: good callout

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: Yes, should improve numerical stability. 

will change `return (1 + x**(-c))**(-d)` to `return np.exp(self._logcdf(x,c,d))`

File: scipy/stats/_continuous_distns.py (hunk 3 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: yes. Will switch `return 1 - (1 + x**(-c))**(-d)` to `return np.exp(self._logsf(x,c,d))`

File: scipy/stats/tests/test_distributions.py (hunk 6 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: Yes, o/w this is the most useless unit test ever :)

**Comment by rlucas7** [comment]
I addressed some of the comments and pushed the changes. The moment calculations/unit tests should be (mostly) ok now. I'm still getting test failures locally on pdf - logpdf roundtrip comparisons. 
I'll try to address that tomorrow evening. Maybe the PR that @pvanmulbregt opened a couple hours ago addresses the numeric errors I'm getting?

File: scipy/stats/tests/test_distributions.py (hunk 5 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: added a comment, these should come from the formula for mean and variance: 

https://edoc.unibas.ch/61230/1/20180305141821_5a9d439d76be5.pdf

page 6 or 8 depending on if you use pdf viewer page numbering or the numbering written on each page of the document. Eqn (7) gives the Kth moment. In the `burr_gen` class `b=1` and `p` in the pdf corresponds to `d` here in the code. The `a` in the pdf corresponds to the `c` here in the code. I'll double check these calculations tomorrow evening.

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (ev-br) [reply scipy/stats/_continuous_distns.py]: Nope, my comment is way more trivial, and is only about whitespace.

File: scipy/stats/tests/test_distributions.py (hunk 5 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: For the `mean_hc` and `variance_hc` I get:
```
>>> sc.beta(3 + 1/5, 1. - 1/5) * 3 
1.4110263183925855
>>> sc.beta(3 + 2/5, 1. - 2/5) * 3  - (sc.beta(3 + 1/5, 1. - 1/5) * 3)**2
0.2287994802619171
```

note there might be some loss in precision in the last few digits displayed. 
Which corresponds to what I would expect based on formulas in the manuscript:

https://edoc.unibas.ch/61230/1/20180305141821_5a9d439d76be5.pdf

and agrees with the output of `_stats` e.g. 

```

```

File: scipy/stats/_continuous_distns.py [hunk 7]
```diff
@@ -757,37 +757,74 @@ class burr_gen(rv_continuous):
 
     .. math::
 
-        f(x, c, d) = c d x^{-c-1} (1+x^{-c})^{-d-1}
+        f(x, c, d) = c d x^{-c - 1} / (1 + x^{-c})^{d - 1}
```
Comment (pvanmulbregt) [inline scipy/stats/_continuous_distns.py]: Should this be `... (1 + x^{-c})^{d + 1}`?  That would match the implementation.

File: scipy/stats/_continuous_distns.py [hunk 8]
```diff
@@ -906,14 +943,45 @@ def _pdf(self, x, c):
...
+             (e3 - 3 * e2 * e1 + 2 * e1**3) / np.sqrt((mu2_if_c)**3),
+             fillvalue=np.nan)
+        g2 = _lazywhere(c > 3.0, (c, e1, e2, e3, e4, mu2_if_c),
+             lambda c, e1, e2, e3, e4, mu2_if_c:
+             ((e4 - 4 * e3 * e1 + 6 * e2 * e1**2 - 3 * e1**4) /
+             mu2_if_c**2) - 3.,
+             fillvalue=np.nan)
+        return mu, mu2, g1, g2
+
+    def _stats(self, c):
```
Comment (pvanmulbregt) [inline scipy/stats/_continuous_distns.py]: There are two implementations of `_stats(self, c)`.  Which one should be kept?

File: scipy/stats/_distr_params.py [hunk 9]
```diff
@@ -10,7 +10,7 @@
     ['beta', (2.3098496451481823, 0.62687954300963677)],
     ['betaprime', (5, 6)],
     ['bradford', (0.29891359763170633,)],
-    ['burr', (10.5, 4.3)],
+    ['burr', (10.0, 5.0)],
```
Comment (pvanmulbregt) [inline scipy/stats/_distr_params.py]: Why change the default parameters?  Do they need to be integers with an integer difference?

Comment (rlucas7) [reply scipy/stats/_distr_params.py]: I did this to make it easier for me to verify via hand calculations, e.g. I can do `\Gamma(9)` but I have difficulty caculating `\Gamma(9.5)` by pen and paper. I can change them back if preferred.

**Comment by pvanmulbregt** [comment]
@rlucas7 I believe that all of this PR may now be incorporated into gh-9983. (Exception: I left the default testing parameters unchanged.) Is there anything missing?

**Comment by rlucas7** [comment]
Thanks for the ping, I’ll take a closer look tonight. 

Sincerely, 

-Lucas Roberts

**Comment by rlucas7** [comment]
Closing in favor of PR #9983

**Pull request closed without merging**